### PR TITLE
[Geo] Gathering Nonconverged solutions in Newton Raphson Erosion Process

### PR DESCRIPTION
**Pull request by Rbravo555**
**📝 Description**
Similarly to #12375, this PR adds a method to activate the gathering of nonconverged solutions during the iterative process of the GeoMechanicsNewtonRaphsonErosionProcessStrategy. 

If the Setup method is not called, no overhead is introduced.

**🆕 Changelog**
- Added a setup method to trigger the storage of the nonconverged solutions
- Added a get method that returns both the matrix of nonconvered solutions and the dofs set

### CODE REVIEW

File: applications/GeoMechanicsApplication/custom_strategies/strategies/geo_mechanics_newton_raphson_erosion_process_strategy.hpp [hunk 1]
```diff
@@ -351,6 +424,10 @@ class GeoMechanicsNewtonRaphsonErosionProcessStrategy
             //    grow = false;
             //}
 
+            if (mStoreNonconvergedSolutionsFlag) {
+                CollectCurrentSolution();
+            }
```
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_strategies/strategies/geo_mechanics_newton_raphson_erosion_process_strategy.hpp]: Just a question to verify my understanding: since the boolean is called `mStoreNonconvergedSolutionsFlag`, should it also be checked if the solution was converged and only collect the current solution with it hasn't? If we only want to save non-converged solutions, I'd propose the following:
```suggestion
            if (mStoreNonconvergedSolutionsFlag && !converged) {
                CollectCurrentSolution();
            }
```

or put it as the `else` after the if-statement of line 431

File: applications/GeoMechanicsApplication/custom_strategies/strategies/geo_mechanics_newton_raphson_erosion_process_strategy.hpp [hunk 2]
```diff
@@ -327,6 +381,25 @@ class GeoMechanicsNewtonRaphsonErosionProcessStrategy
...
+     * @brief Collects the current solution vector for the degrees of freedom (DOFs).
+     * @details This method retrieves the current solution values for the provided DOF set.
+     * The solution vector will be resized to match the size of the DOF set if necessary,
+     * and will be filled with the solution values corresponding to each DOF. Each value is accessed
+     * using the equation ID associated with each DOF.
+     */
+    void CollectCurrentSolution() {
+        auto& rDofSet = this->GetBuilderAndSolver()->GetDofSet();
+        Vector this_solution(rDofSet.size());
+        block_for_each(rDofSet, [&](const auto& r_dof) {
```
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_strategies/strategies/geo_mechanics_newton_raphson_erosion_process_strategy.hpp]: Could you explicitly capture the variables needed in the lambda expression here, instead of capturing everything implicitly?
```suggestion
        block_for_each(rDofSet, [&this_solution](const auto& r_dof) {
```

File: applications/GeoMechanicsApplication/custom_strategies/strategies/geo_mechanics_newton_raphson_erosion_process_strategy.hpp [hunk 3]
```diff
@@ -140,6 +181,19 @@ class GeoMechanicsNewtonRaphsonErosionProcessStrategy
             KRATOS_ERROR_IF_NOT(converged) << "Groundwater flow calculation failed to converge." << std::endl;
         }
 
+        if (mStoreNonconvergedSolutionsFlag) {
+            CollectCurrentSolution();
+            auto& r_dof_set = this->GetBuilderAndSolver()->GetDofSet();
+            mNonconvergedSolutionsMatrix = Matrix( r_dof_set.size(), mNonconvergedSolutions.size() );
+            for (std::size_t i = 0; i < mNonconvergedSolutions.size(); ++i) {
+                block_for_each(r_dof_set, [&](const auto& r_dof) {
```
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_strategies/strategies/geo_mechanics_newton_raphson_erosion_process_strategy.hpp]: Could you explicitly capture the necessary variables for the lambda function? I think this should work:
```suggestion
                block_for_each(r_dof_set, [this, &mNonconvergedSolutions, i](const auto& r_dof) {
```

File: applications/GeoMechanicsApplication/custom_strategies/strategies/geo_mechanics_newton_raphson_erosion_process_strategy.hpp [hunk 4]
```diff
@@ -78,6 +78,47 @@ class GeoMechanicsNewtonRaphsonErosionProcessStrategy
...
+     * @brief This matrix stores the non-converged solutions
+     * @details The matrix is structured such that each column represents the solution vector at a specific non-converged iteration.
+     */
+    Matrix mNonconvergedSolutionsMatrix;
+
+    /**
+     * @brief Flag indicating whether to store non-converged solutions
+     * @details Only when set to true (by calling the SetUpNonconvergedSolutionsGathering method) will the non-converged solutions at each iteration be stored.
+     */
+    bool mStoreNonconvergedSolutionsFlag = false;
```
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_strategies/strategies/geo_mechanics_newton_raphson_erosion_process_strategy.hpp]: Did you consider making these members private? If I see this PR, we get access to the relevant variables through the new public `GetNonconvergedSolutions` function, so I think the member variables can be private.

File: applications/GeoMechanicsApplication/custom_strategies/strategies/geo_mechanics_newton_raphson_erosion_process_strategy.hpp [hunk 5]
```diff
@@ -78,6 +78,47 @@ class GeoMechanicsNewtonRaphsonErosionProcessStrategy
...
+
+
+    /**
+     * @brief Gets the matrix of non-converged solutions and the DOF set
+     * @return A tuple containing the matrix of non-converged solutions and the DOF set
+     * @details This method returns a tuple where the first element is a matrix of non-converged solutions and the second element is the corresponding DOF set.
+     */
+    std::tuple<Matrix, DofsArrayType> GetNonconvergedSolutions(){
```
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_strategies/strategies/geo_mechanics_newton_raphson_erosion_process_strategy.hpp]: Do you think it is feasible to add a test for this new behavior? If it is not too much work, that would certainly be worthwhile in my opinion (such that we are sure that this functionality keeps working as it does when the `mStoreNonconvergedSolutionsFlag` is set to true

**Review by rfaasse — COMMENTED** [review]
This is a clear PR and some nice functionality that might also be useful for other processes that depend on non-convergence (like stability analysis)! 

I have a few suggestions to remove a couple of code smells and some questions to validate my understanding, but from my perspective this looks almost ready to go.

File: applications/GeoMechanicsApplication/custom_strategies/strategies/geo_mechanics_newton_raphson_erosion_process_strategy.hpp
Comment (avdg81) [inline applications/GeoMechanicsApplication/custom_strategies/strategies/geo_mechanics_newton_raphson_erosion_process_strategy.hpp:1]: It appears that some code here has been copied from class `GeoMechanicsNewtonRaphsonStrategy`, which results in hiding some member functions of class template `GeoMechanicsNewtonRaphsonErosionProcessStrategy`. For instance, member function `GetNonconvergedSolutions()` (at line 106 of file `geo_mechanics_newton_raphson_erosion_process_strategy.hpp`) is identical to its counterpart in the base class (see line 903 of file `residualbased_newton_raphson_strategy.h`). So my suggestion would be to not repeat it, but to use this base functionality. A similar reasoning applies to most (but not all) new data members that have been added to `GeoMechanicsNewtonRaphsonErosionProcessStrategy`.

In short, I would strongly recommend to revisit the changes that have been made and to see which parts of the base class functionality can be reused, to avoid repeating certain functionality. Please feel free to contact me if it is unclear how to do that, or when you run into technical difficulties. I'm here to help you, if needed. Thanks.

File: applications/GeoMechanicsApplication/custom_strategies/strategies/geo_mechanics_newton_raphson_erosion_process_strategy.hpp (hunk 3 above)
Comment (roigcarlo) [reply applications/GeoMechanicsApplication/custom_strategies/strategies/geo_mechanics_newton_raphson_erosion_process_strategy.hpp]: probably this will give problems with the member, since we can use c++14 I would change to `[&rNonConvergedSolutions = this->mNonconvergedSolutions, i]` or just capture `this` but without referencing `mNonconvergedSolutions` in the capture clause.

File: applications/GeoMechanicsApplication/custom_strategies/strategies/geo_mechanics_newton_raphson_erosion_process_strategy.hpp
Comment (Rbravo555) [reply applications/GeoMechanicsApplication/custom_strategies/strategies/geo_mechanics_newton_raphson_erosion_process_strategy.hpp:1]: I had indeed repeated some code from the base class `ResidualBasedNewtonRaphsonStrategy.` Now, I have reused the parts I could, but I still had to define the function `CollectCurrentSolution`  and the three variables `mNonconvergedSolutions`, `mNonconvergedSolutionsMatrix`, and `mStoreNonconvergedSolutionsFlag` .

File: applications/GeoMechanicsApplication/custom_strategies/strategies/geo_mechanics_newton_raphson_erosion_process_strategy.hpp (hunk 5 above)
Comment (Rbravo555) [reply applications/GeoMechanicsApplication/custom_strategies/strategies/geo_mechanics_newton_raphson_erosion_process_strategy.hpp]: I have added a python test for the nonconverged solution gathering. I think this way is easier than implementing a cpp test as in the `test_GeoMechanisNewtonRaphsonErosionProcess.cpp`.  I this a cpp test would involve touching  `dgeoflow.cpp` to access the solving strategy and activate the flag.

Comment (Rbravo555) [reply applications/GeoMechanicsApplication/custom_strategies/strategies/geo_mechanics_newton_raphson_erosion_process_strategy.hpp]: The test however has not been included in the small or nightly test suite yet
